sympy/printing/tests/test_c.py

```diff
@@ -578,7 +578,7 @@ def test_ccode_Assignment():
 
 def test_ccode_For():
     f = For(x, Range(0, 10, 2), [aug_assign(y, '*', x)])
-    assert ccode(f) == ("for (x = 0; x < 10; x += 2) {\n"
+    assert ccode(f) == ("for (int x=0; x<10; x+=2){\n"
```

Comment:
hmmm, wouldn't this also be a breaking change?

Reply:
I'm not sure that it will break something since this example

```
x, y = symbols('x, y')
expr = For(x, Range(0, 10, 2), [aug_assign(y, '*', x)])
name_expr = ("test", [expr])

result = codegen(name_expr, "c", "test", header=False, empty=False)
source = result[0][1]
print(source)
```
is broken and give
```
#include "test.h"
#include <math.h>
double test(double x, double y) {
   double test_result;
   test_result = for (x = 0; x < 10; x += 2) {
      y *= x;
   };
   return test_result;
}
```

It depends if we want to keep a common format or not. And there are a number of things in the codegen where there are unfortunately implementations that are not consistent at the moment...

Reply:
Well, I've seen that many projects use templates for code-generation (I for sure have done so quite extensively). That means `codegen(...)` (from `sympy.utilities.codegen`) is never really involved at all, but instead only functions in `sympy.printing` and classes in `sympy.codegen`.

It is a bit unfortunate that we haven't reimplemented all the functionality from `sympy.utilities.codegen` module (which predates the aforementioned). But as you are well aware, `sympy.utilities.codegen` has many features which have not yet been added to the younger `sympy.codegen` submodule.

Therefore, we cannot defend breaking a test in say, `.printing.tests.test_c` based on how `.utilities.codegen` operates.

Reply:
@bjodah I understand. I have no idea how many people use codegen in their code. But I would be very happy to have some links to projects if you have any.

The test is not broken if you specify the standard like

```
def test_ccode_For():
    f = For(x, Range(0, 10, 2), [aug_assign(y, '*', x)])
    assert ccode(f, standard='C89') == ("for (x=0; x<10; x+=2){\n"
                        "   y *= x;\n"
                        "}")
    assert ccode(f, standard='C99') == ("for (int x=0; x<10; x+=2){\n"
                        "   y *= x;\n"
                        "}")
```

Is it a good option to specify the standard?

Reply:
So this is, admittedly, a silly example, but still:
```python
from sympy import symbols, ccode, Range                                                                                                                                                                                                                                       
from sympy.codegen import aug_assign, For                                                                                                                                                                                                                                     
i, y, argc, cexpr = symbols('i y argc atof(argv[i])')                                                                                                                                                                                                                         
loop = For(i, Range(1, argc, 1), [aug_assign(y, '*', cexpr)])                                                                                                                                                                                                                 
print("""                                                                                                                                                                                                                                                                     
#include <stdio.h>                                                                                                                                                                                                                                                            
#include <stdlib.h>                                                                                                                                                                                                                                                           
int main(int argc, char *argv[]) {                                                                                                                                                                                                                                            
    int i;                                                                                                                                                                                                                                                                    
    double y = 1;                                                                                                                                                                                                                                                             
    %(loop)s                                                                                                                                                                                                                                                                  
    printf("%%e\\n", y);                                                                                                                                                                                                                                                      
    return 0;                                                                                                                                                                                                                                                                 
}                                                                                                                                                                                                                                                                             
""" % dict(loop=ccode(loop)))
```
and on the command line:
```
$ python3 silly_codegen_example.py >a.c && gcc -Wall -Wextra a.c && ./a.out 2 3 4                                                                                                                                                              
2.400000e+01
```
So the test I was commenting on, ensures that code like that keeps working. If we change `ccode` to emit "int" by default, then users would get compilation errors (redeclaration of the loop variable):
```
a.c: In function ‘main’:                                                                                                                                                                                                                                                      
a.c:7:14: error: redeclaration of ‘i’ with no linkage                                                                                                                                                                                                                         
     for (int i = 1; i < argc; i += 1) {                                                                                                                                                                                                                                      
              ^                                                                                                                                                                                                                                                               
```

We should compile a list of packages using sympy for codegen, it's a nice idea. I remember using advanced options in github search a couple of years ago (using "sympy" and "ccode" and clever negative options for filtering out forks of the SymPy repo). Unfortunately I forgot exactly what that query looked like, neither can I find any notes of what projects I found.